src/py21cmfast/wrapper.py

```diff
@@ -3219,3 +3223,308 @@ def calibrate_photon_cons(
+
+
+class KszConstants:
+    """Constants used for kSZ calculation."""
+
+    def __init__(
+        self,
+        HII_DIM,
+        BOX_LEN,
+        hlittle,
+        OMb,
+        red_dist,
+        redshift_start,
+        DA_zstart,
+        Y_He,
+        Zreion_HeII
+        #    mean_taue_curr_z
+    ):
+        RHOb_cgs = (
+            3.0
+            * (hlittle * 3.2407e-18) ** 2
+            / (8.0 * np.pi * constants.G.cgs.value)
+            * OMb
+            / constants.m_p.cgs.value
+        )  # pcm^-3 at z=0
+        self.He_No = (
+            RHOb_cgs * self.Y_He / 4.0
+        )  # current helium number density estimate
+        self.N_b0 = RHOb_cgs * (
+            1 - 0.75 * self.Y_He
+        )  # present-day baryon num density, H + He
+        self.dR = BOX_LEN / HII_DIM
+        self.CMperMPC = constants.kpc.cgs.value * 1e3
+        self.A = self.N_b0 * constants.sigma_T.cgs.value * self.dR * self.CMperMPC
+        self.HII_DIM = HII_DIM
+        self.BOX_LEN = BOX_LEN
+        self.red_dist = red_dist
+        self.redshift_start = redshift_start
+        self.DA_zstart = DA_zstart
+        self.Zreion_HeII = Zreion_HeII
+
+
+def drdz(z, cosmo_params):
+    """Comiving distance (in cm) per unit redshift."""
+    return (1.0 + z) * constants.c.cgs.value * dtdz(z, cosmo_params)
+
+
+def dtdz(z, cosmo_params):
+    """The value of dt/dz at the redshift parameter z."""
+    OMm = cosmo_params.OMm
+    OMl = 1 - OMm
+    hlittle = cosmo_params.hlittle
+    Ho = hlittle * 3.2407e-18  # /* Hubble parameter[s^-1] at z=0 */
+
+    x = np.sqrt(OMl / OMm) * (1 + z) ** (-3.0 / 2.0)
+    dxdz = np.sqrt(OMl / OMm) * (1 + z) ** (-5.0 / 2.0) * (-3.0 / 2.0)
+    const1 = 2 * np.sqrt(1 + OMm / OMl) / (3.0 * Ho)
+    numer = dxdz * (1 + x * (x ** 2 + 1) ** (-0.5))
+    denom = x + np.sqrt(x ** 2 + 1)
+    return const1 * numer / denom
+
+
+class TaueParams:
+    """Class for stuff needed in the optical depth calculation."""
+
+    def __init__(self, z, xH, length, cosmo_params):
+        self.z = z
+        self.i = xH
+        self.length = length
+        self.cosmo_params = cosmo_params
+
+
+def _dtau_e_dz(z_en, z, xH, length, cosmo_params):
+    """Calculation of optical depth per redshift."""
+    p = TaueParams(z, xH, length, cosmo_params)
+    if (p.length == 0) or not (p.z):
+        return (1 + z_en) * (1 + z_en) * drdz(z_en, cosmo_params)
+    else:
+        if p.z[0] > z_en:
+            return (1 + z_en) * (1 + z_en) * drdz(z_en, cosmo_params)
+        if all(np.array(p.z) < z_en):
+            return 0
+        else:
+            i = [i for i, x in enumerate(np.array(p.z) > z_en) if x][0]
+        xH_inter = p.xH[i - 1] + (p.xH[i] - p.xH[i - 1]) / (p.z[i] - p.z[i - 1]) * (
+            z_en - p.z[i - 1]
+        )
+        xi = 1.0 - xH_inter
+        if xi < 0:
+            xi = 0
+        if xi > 1:
+            xi = 1
+        return xi * (1 + z_en) * (1 + z_en) * drdz(z_en, cosmo_params)
+
+
+def _tau_e(
```

Comment:
we already have a C function ('src/UsefulFunctions.c') to calcualtion this. please replace the following with existing C code

Reply:
I'm using compute_tau function from wrapper.py which calls the C function!